OpenSim/Actuators/Test/testMuscles.cpp

```diff
@@ -1108,10 +929,95 @@ void testDelp1990Muscle()
         x0, 
         act0, 
         &motion, 
-        &control, 
-        IntegrationAccuracy,
-        SimulationTest,
-        SimulationTestTolerance,
+        &control,
         false);
+}
+
+
+
+void testMuscleEquilibriumSolve(const Model& model, const Storage& statesStore)
+{
+    // Get the muscle to test
+    const Muscle& muscle = model.getMuscles()[0];
+
+    if (dynamic_cast<const Millard2012AccelerationMuscle*>(&muscle)) {
+        // Millard2012AccelerationMuscle is not an Equilibrium-based Muscle
+        return;
+    }
+
+    // Load input data as StatesTrajectory used to perform the Analysis
+    auto statesTraj = StatesTrajectory::createFromStatesStorage(
+        model, statesStore, true, false);
+    size_t nstates = statesTraj.getSize();
 
+    // muscle active, passive, total muscle and tendon force quantities
+    double af, pf, mf, tf, nfl, fv;
+    af = pf = mf = tf = nfl = fv = SimTK::NaN;
+
+    // Tolerance for muscle equilibrium solution 
+    const double equilTol = muscle.getMaxIsometricForce()*SimTK::SqrtEps;
+
+    // The maximum acceptable change in force between two contiguous states
+    const double maxDelta = muscle.getMaxIsometricForce() / 2;
+
+    const int N = 20;
+    const double minActivation = muscle.getMinControl();
+    const double dAct = (1.0 - minActivation) / N;
+    double activation = 0;
+
+    SimTK::State s = model.getWorkingState();
+    // Independently compute the active fiber force at every state
+    for (int i = 0; i < nstates; ++i) {
+        s = statesTraj[i];
+
+        // test a full sweep of default activations at each state
+        for (int j = 0; j <= N; ++j) {
+            activation = minActivation + j*dAct;
+            muscle.setActivation(s, activation);
+
+            try {
+                muscle.computeEquilibrium(s);
+            }
+            catch (const MuscleCannotEquilibrate& x) {
+                // Write out the muscle equilibrium error as a function of
+                // fiber lengths.
+                const Thelen2003Muscle* thelen =
+                    dynamic_cast<const Thelen2003Muscle*>(&muscle);
+                if (thelen) {
+                    thelen->printCurveToCSVFile(
+                        Thelen2003Muscle::CurveType::FiberForceVelocity, "");
+                    reportTendonAndFiberForcesAcrossFiberLengths(*thelen, s);
+                }
+                else {
```

Comment:
No need to change anything, but this would be cleaner as:

```cpp
if (const auto* thelen = dynamic_cast...) {
    ...
} else if (const auto* millard = dynamic_cast...) {
    ...
}
```